File: pkg/material/glass.go
package material

import (
	"math"
	"math/rand"

	"sif/gogs/eric/relray/pkg/geometry"
	"sif/gogs/eric/relray/pkg/spectrum"
	"sif/gogs/eric/relray/pkg/vec"
)

// Glass is a dielectric material with refraction and Fresnel reflection.
// Uses Schlick's approximation for the Fresnel term.
type Glass struct {
	// IOR is the index of refraction (e.g., 1.5 for typical glass).
	IOR float64
	// Tint is the spectral transmittance of the glass (1.0 = perfectly clear).
	// Values < 1 at certain wavelengths create colored glass.
	Tint spectrum.SPD
}

func (g *Glass) Scatter(inDir vec.Vec3, hit geometry.Hit, rng *rand.Rand) ScatterResult {
	// Determine if we're entering or exiting the glass
	var etaRatio float64 // eta_incident / eta_transmitted
	if hit.FrontFace {
		etaRatio = 1.0 / g.IOR // air → glass
	} else {
		etaRatio = g.IOR // glass → air
	}

	unitDir := inDir.Normalize()
	cosI := math.Min(-unitDir.Dot(hit.Normal), 1.0)
	sin2T := etaRatio * etaRatio * (1.0 - cosI*cosI)

	// Total internal reflection check
	if sin2T > 1.0 {
		reflected := unitDir.Reflect(hit.Normal)
		return ScatterResult{
			Scattered:   true,
			OutDir:      reflected.Normalize(),
			Reflectance: g.Tint,
		}
	}

	// Schlick's approximation for Fresnel reflectance
	reflectance := schlick(cosI, etaRatio)

	// Probabilistically choose reflection vs refraction
	if rng.Float64() < reflectance {
		reflected := unitDir.Reflect(hit.Normal)
		return ScatterResult{
			Scattered:   true,
			OutDir:      reflected.Normalize(),
			Reflectance: g.Tint,
		}
	}

	// Refract
	refracted := refract(unitDir, hit.Normal, etaRatio)
	return ScatterResult{
		Scattered:   true,
		OutDir:      refracted.Normalize(),
		Reflectance: g.Tint,
	}
}

func (g *Glass) Emitted(hit geometry.Hit) spectrum.SPD {
	return spectrum.SPD{}
}

// schlick computes Schlick's approximation to the Fresnel reflectance.
func schlick(cosine float64, etaRatio float64) float64 {
	r0 := (1 - etaRatio) / (1 + etaRatio)
	r0 = r0 * r0
	return r0 + (1-r0)*math.Pow(1-cosine, 5)
}

// refract computes the refracted direction using Snell's law.
func refract(uv, n vec.Vec3, etaRatio float64) vec.Vec3 {
	cosTheta := math.Min(-uv.Dot(n), 1.0)
	rPerp := uv.Add(n.Scale(cosTheta)).Scale(etaRatio)
	rParallel := n.Scale(-math.Sqrt(math.Abs(1.0 - rPerp.LengthSq())))
	return rPerp.Add(rParallel)
}

File: rust/src/material/glass.rs
use crate::geometry::Hit;
use crate::spectrum::Spd;
use crate::vec::Vec3;
use rand::Rng;
use rand::rngs::SmallRng;

use super::{Material, ScatterResult};

/// Dielectric material with refraction and Fresnel reflection.
/// Uses Schlick's approximation for the Fresnel term.
pub struct Glass {
    /// Index of refraction (e.g., 1.5 for typical glass).
    pub ior: f64,
    /// Spectral transmittance of the glass (1.0 = perfectly clear).
    /// Values < 1 at certain wavelengths create colored glass.
    pub tint: Spd,
}

impl Material for Glass {
    fn scatter(&self, in_dir: Vec3, hit: &Hit, rng: &mut SmallRng) -> ScatterResult {
        // Determine if we're entering or exiting the glass
        let eta_ratio = if hit.front_face {
            1.0 / self.ior // air -> glass
        } else {
            self.ior // glass -> air
        };

        let unit_dir = in_dir.normalize();
        let cos_i = (-unit_dir.dot(hit.normal)).min(1.0);
        let sin2_t = eta_ratio * eta_ratio * (1.0 - cos_i * cos_i);

        // Total internal reflection check
        if sin2_t > 1.0 {
            let reflected = unit_dir.reflect(hit.normal);
            return ScatterResult {
                scattered: true,
                out_dir: reflected.normalize(),
                reflectance: self.tint,
            };
        }

        // Schlick's approximation for Fresnel reflectance
        let reflectance = schlick(cos_i, eta_ratio);

        // Probabilistically choose reflection vs refraction
        if rng.random_range(0.0_f64..1.0) < reflectance {
            let reflected = unit_dir.reflect(hit.normal);
            return ScatterResult {
                scattered: true,
                out_dir: reflected.normalize(),
                reflectance: self.tint,
            };
        }

        // Refract
        let refracted = refract(unit_dir, hit.normal, eta_ratio);
        ScatterResult {
            scattered: true,
            out_dir: refracted.normalize(),
            reflectance: self.tint,
        }
    }

    fn emitted(&self, _hit: &Hit) -> Spd {
        Spd::default()
    }
}

/// Schlick's approximation to the Fresnel reflectance.
fn schlick(cosine: f64, eta_ratio: f64) -> f64 {
    let r0 = (1.0 - eta_ratio) / (1.0 + eta_ratio);
    let r0 = r0 * r0;
    r0 + (1.0 - r0) * (1.0 - cosine).powi(5)
}

/// Computes the refracted direction using Snell's law.
fn refract(uv: Vec3, n: Vec3, eta_ratio: f64) -> Vec3 {
    let cos_theta = (-uv.dot(n)).min(1.0);
    let r_perp = (uv + n * cos_theta) * eta_ratio;
    let r_parallel = n * -(1.0 - r_perp.length_sq()).abs().sqrt();
    r_perp + r_parallel
}
